flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java

```diff
@@ -0,0 +1,422 @@
+	 *                            path is exclusive for all state handles managed by this
+	 *                            instance, e.g. <code>client.usingNamespace("/stateHandles")</code>
+	 * @param stateHandleProvider The state handle provider for the state
+	 */
+	public ZooKeeperStateHandleStore(
+			CuratorFramework client,
+			StateHandleProvider<T> stateHandleProvider) {
+
+		this.client = checkNotNull(client, "Curator client");
+		this.stateHandleProvider = checkNotNull(stateHandleProvider, "State handle provider");
+	}
+
+	/**
+	 * Creates a state handle and stores it in ZooKeeper with create mode {@link
+	 * CreateMode#PERSISTENT}.
+	 *
+	 * @see #add(String, Serializable, CreateMode)
+	 */
+	public ZooKeeperStateHandle<T> add(String pathInZooKeeper, T state) throws Exception {
+		return add(pathInZooKeeper, state, CreateMode.PERSISTENT);
+	}
+
+	/**
+	 * Creates a state handle and stores it in ZooKeeper.
+	 *
+	 * <p><strong>Important</strong>: This will <em>not</em> store the actual state in
+	 * ZooKeeper, but create a state handle and store it in ZooKeeper. This level of indirection
+	 * makes sure that data in ZooKeeper is small.
+	 *
+	 * @param pathInZooKeeper Destination path in ZooKeeper (expected to *not* exist yet and
+	 *                        start with a '/')
+	 * @param state           State to be added
+	 * @param createMode      The create mode for the new path in ZooKeeper
+	 * @return Created {@link ZooKeeperStateHandle}
+	 * @throws Exception
+	 */
+	public ZooKeeperStateHandle<T> add(String pathInZooKeeper, T state, CreateMode createMode) throws Exception {
+		checkNotNull(pathInZooKeeper, "Path in ZooKeeper");
+		checkNotNull(state, "State");
+
+		// Create the state handle. Nothing persisted yet.
+		StateHandle<T> stateHandle = stateHandleProvider.createStateHandle(state);
+
+		boolean success = false;
+
+		try {
+			// Serialize the state handle. This writes the state to the backend.
+			byte[] serializedStateHandle = InstantiationUtil.serializeObject(stateHandle);
+
+			// Write state handle (not the actual state) to ZooKeeper. This is expected to be
+			// smaller than the state itself. This level of indirection makes sure that data in
+			// ZooKeeper is small, because ZooKeeper is designed for data in the KB range, but
+			// the state can be larger.
+			client
+					.create()
+					.withMode(createMode)
+					.forPath(pathInZooKeeper, serializedStateHandle);
+
+			success = true;
+
+			return new ZooKeeperStateHandle<>(stateHandle, pathInZooKeeper);
+		}
+		finally {
+			if (!success) {
+				// Cleanup the state handle if it was not written to ZooKeeper.
+				if (stateHandle != null) {
+					stateHandle.discardState();
+				}
+			}
+		}
+	}
+
+	/**
+	 * Replaces a state handle in ZooKeeper and discards the old state handle.
+	 *
+	 * <p><strong>Important</strong>: This method will only discard the state handle and not the
+	 * state itself. Don't forget to run custom cleanup code of the state, if necessary.
```

Comment:
There is actually one word missing in the comment: the *old* state handle. This is a copy & paste error from the remove method.

This class is used by both the `ZooKeeperSubmittedJobGraphs` and `ZooKeeperCompletedCheckpoints`. For `ZooKeeperSubmittedJobGraphs` you can just discard the state handle (the above comment does not apply), because there is no custom logic cleanup logic for the job graph. But for a `SuccessfulCheckpoint`, there is required cleanup, which would be skipped if you directly discard the state handle.

The clean up logic for the SuccessfulCheckpoint is this (where StateForTask contains further state handles):
```java
	public void discard(ClassLoader userClassLoader) {
		for(StateForTask state: states){
			state.discard(userClassLoader);
		}
		states.clear();
	}
``` 

If you create a state handle for a SuccessfulCheckpoint you and store it in ZooKeeper you still need to call this to make sure that all `states` are discarded. Otherwise we just remove the state handle to the checkpoint:

```
    [s1,  s2,  s3]
      \    |    /
  SuccessfulCheckpoint
           | (create state handle and store in ZK)
        [ s4 ]
```
We store `s4` in ZooKeeper. When we now discard `s4` (before replacing), we have to make sure to discard `s1, s2, s3` (the state for the tasks) as well. If we don't do this, the old state handle will still exist.

What can we do to make this clearer? Or am I wrong here? If you didn't get it, I doubt that others will have an easier time.